crates/extensions/tedge_mqtt_bridge/src/test_helpers.rs

```diff
@@ -0,0 +1,412 @@
+        Ok(Event::Incoming(Incoming::SubAck(SubAck {
+            pkid: 1,
+            return_codes: vec![SubscribeReasonCode::Success(QoS::AtLeastOnce)],
+        })))
+    };
+    (publish($msg:expr)) => {
+        Ok(Event::Incoming(Incoming::Publish($msg.clone())))
+    };
+    (puback($pkid:expr)) => {
+        Ok(Event::Incoming(Incoming::PubAck(PubAck { pkid: $pkid })))
+    };
+    (network_error) => {
+        Err(ConnectionError::NetworkTimeout)
+    };
+}
+
+#[macro_export]
+macro_rules! out {
+    (publish($pkid:expr)) => {
+        Ok(Event::Outgoing(Outgoing::Publish($pkid)))
+    };
+}
+
+pub type EventRes = Result<Event, rumqttc::ConnectionError>;
+
+#[async_trait::async_trait]
+/// Encapsulates the logic for waiting for all messages to finish processing
+pub trait AllProcessed {
+    async fn all_processed(&self) -> anyhow::Result<()>;
+}
+
+#[derive(Default, Debug, Clone)]
+/// A fixed stream of events
+pub struct FixedEventStream {
+    events: Arc<Mutex<VecDeque<EventRes>>>,
+}
+
+impl FixedEventStream {
+    fn next_event(&self) -> Option<EventRes> {
+        self.events.lock().unwrap().pop_front()
+    }
+}
+
+impl<I: Into<VecDeque<EventRes>>> From<I> for FixedEventStream {
+    fn from(value: I) -> Self {
+        Self {
+            events: Arc::new(Mutex::new(value.into())),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl MqttEvents for FixedEventStream {
+    async fn poll(&mut self) -> Result<Event, ConnectionError> {
+        if let Some(event) = self.next_event() {
+            event
+        } else {
+            pending().await
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl AllProcessed for FixedEventStream {
+    async fn all_processed(&self) -> anyhow::Result<()> {
+        let timeout = Duration::from_secs(5);
+        let start = Instant::now();
+        while !self.events.lock().unwrap().is_empty() {
+            if start.elapsed() > timeout {
+                bail!("Timed out waiting for event emitter to be fully consumed. Unconsumed events were {:?}", self.events.lock().unwrap())
+            }
+
+            tokio::task::yield_now().await;
+        }
+        Ok(())
+    }
+}
+
+pub struct EventsPolled;
+
+#[derive(Clone)]
+/// An [MqttClient] implementation that blocks on [MqttClient::subscribe_many]
+///
+/// This is used to verify that [rumqttc::EventLoop::poll] is called
+pub struct BlockingSubscribeClient {
+    pub rx: Arc<TokioMutex<Option<oneshot::Receiver<EventsPolled>>>>,
+}
+
+impl BlockingSubscribeClient {
+    pub fn new(rx: oneshot::Receiver<EventsPolled>) -> Self {
+        Self {
+            rx: Arc::new(TokioMutex::new(Some(rx))),
+        }
+    }
+}
+
+#[derive(Clone)]
+/// Designed to be used with [BlockingSubscribeClient]
+///
+/// Polling this will unblock the call to [BlockingSubscribeClient::subscribe_many]
+pub struct UnblockingEventStream {
+    pub events: FixedEventStream,
+    pub tx: Arc<TokioMutex<Option<oneshot::Sender<EventsPolled>>>>,
+}
```

Comment:
Since these structs are used exclusively for one single test, wouldn't it be better to keep these structs along with that test? Unless you feel that these can still be used for more tests. 

Initially I had a similar impression of the `ChannelClient` and `ChannelEvents` duo as well, but I see more scope for reuse for them compared to these.